src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,133 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[10]; // Ініціалізація масиву з початковим розміром
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            resize(); // Якщо масив заповнений, збільшуємо розмір
+        }
+        elements[size++] = value; // Додаємо елемент і збільшуємо розмір
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+
+        if (size == elements.length) {
+            resize(); // Якщо масив заповнений, збільшуємо розмір
+        }
 
+        // Переміщаємо елементи вправо від індексу до кінця списку
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+
+        // Додаємо новий елемент
+        elements[index] = value;
+        size++; // Збільшуємо розмір
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        // Перебираємо всі елементи переданого списку і додаємо їх до нашого
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i)); // Додаємо кожен елемент в наш список
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index
+                    + ". Valid range: 0 to " + (size - 1));
+        }
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
+        // Перевірка, чи індекс знаходиться в межах
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
 
+        // Заміняємо елемент на новий
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index
```

Comment:
The index validation logic is repeated here. Refactor this logic into a separate method to adhere to the DRY principle.